From 3415d4354010eeccb1add438346b6f4a6558b967 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Mon, 29 May 2023 11:41:07 +0530
Subject: [PATCH] ath12k: fix memory leak caused by thermal debug register

wiphy_free deallocates rdev via put_device which internally uses
kref_count of kobj. This kref_count gets incremented during device_add and
decremented during device_del. During rmmod, wiphy_free expectes this
kref_count to be zero for deallocating rdev.

Ath12k driver uses devm_hwmon_device_register_with_groups for thermal
register by passing hw->wiphy->dev which is the child of pci->dev. So the
kref_count incremented for the hw->wiphy->dev is not getting decremented
properly during device_del which causes stale value present in the
kref_count. Because of the positive value in kref_count, put_device failed
to call kfree of rdev which causes memory leaks.

Fix it by using hwmon_device_register_with_groups instead of devm_hwmon_device_register_with_groups
and free it using hwmon_device_unregiste during unregister.

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/thermal.c | 17 +++++++++++------
 drivers/net/wireless/ath/ath12k/thermal.h |  1 +
 2 files changed, 12 insertions(+), 6 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/thermal.c b/drivers/net/wireless/ath/ath12k/thermal.c
index e29a488..2cf191a 100644
--- a/drivers/net/wireless/ath/ath12k/thermal.c
+++ b/drivers/net/wireless/ath/ath12k/thermal.c
@@ -158,7 +158,6 @@ int ath12k_thermal_set_throttling(struct ath12k *ar, u32 throttle_state)
 int ath12k_thermal_register(struct ath12k_base *ab)
 {
 	struct thermal_cooling_device *cdev;
-	struct device *hwmon_dev;
 	struct ath12k *ar;
 	struct ath12k_pdev *pdev;
 	struct ieee80211_hw *hw;
@@ -199,12 +198,13 @@ int ath12k_thermal_register(struct ath12k_base *ab)
 		if (!IS_REACHABLE(CONFIG_HWMON))
 			return 0;
 
-		hwmon_dev = devm_hwmon_device_register_with_groups(&hw->wiphy->dev,
-								   "ath12k_hwmon", ar,
-								   ath12k_hwmon_groups);
-		if (IS_ERR(hwmon_dev)) {
+		ar->thermal.hwmon_dev = hwmon_device_register_with_groups(&hw->wiphy->dev,
+									  "ath12k_hwmon", ar,
+									  ath12k_hwmon_groups);
+		if (IS_ERR(ar->thermal.hwmon_dev)) {
 			ath12k_err(ar->ab, "failed to register hwmon device: %ld\n",
-				   PTR_ERR(hwmon_dev));
+				   PTR_ERR(ar->thermal.hwmon_dev));
+			ar->thermal.hwmon_dev = NULL;
 			ret = -EINVAL;
 			goto err_thermal_destroy;
 		}
@@ -237,6 +237,11 @@ void ath12k_thermal_unregister(struct ath12k_base *ab)
 		snprintf(pdev_name, sizeof(pdev_name), "%s%d", "cooling_device",
 			 ar->link_idx);
 
+		if (ar->thermal.hwmon_dev) {
+			hwmon_device_unregister(ar->thermal.hwmon_dev);
+			ar->thermal.hwmon_dev = NULL;
+		}
+
 		sysfs_remove_link(&hw->wiphy->dev.kobj, pdev_name);
 		thermal_cooling_device_unregister(ar->thermal.cdev);
 	}
diff --git a/drivers/net/wireless/ath/ath12k/thermal.h b/drivers/net/wireless/ath/ath12k/thermal.h
index 8ae3f84..c940b12 100644
--- a/drivers/net/wireless/ath/ath12k/thermal.h
+++ b/drivers/net/wireless/ath/ath12k/thermal.h
@@ -17,6 +17,7 @@
 struct ath12k_thermal {
 	struct thermal_cooling_device *cdev;
 	struct completion wmi_sync;
+	struct device *hwmon_dev;
 
 	/* protected by conf_mutex */
 	u32 throttle_state;
-- 
2.17.1

